### PR TITLE
PoS Block Producer: Data Structure

### PR DESCRIPTION
**Pull request by AeonSw4n**

### CODE REVIEW

**Comment by AeonSw4n** [comment]
Current dependencies on/for this PR:
* feature/proof-of-stake
  * **PR #578** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/578" target="_blank">Graphite</a> 
    * **PR #671** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/671" target="_blank">Graphite</a> 
      * **PR #676** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/676" target="_blank">Graphite</a> 
        * **PR #700** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/700" target="_blank">Graphite</a> 
          * **PR #701** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/701" target="_blank">Graphite</a> 
            * **PR #702** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/702" target="_blank">Graphite</a> 
              * **PR #703** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/703" target="_blank">Graphite</a> 
                * **PR #704** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/704" target="_blank">Graphite</a>  👈
        * **PR #689** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/689" target="_blank">Graphite</a> 
      * **PR #673** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/673" target="_blank">Graphite</a> 
      * **PR #672** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/672" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/deso-protocol/core/704?utm_source=stack-comment).

**Comment by AeonSw4n** [comment]
Current dependencies on/for this PR:
* feature/proof-of-stake
  * **PR #578** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/578" target="_blank">Graphite</a> 
    * **PR #671** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/671" target="_blank">Graphite</a> 
      * **PR #676** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/676" target="_blank">Graphite</a> 
        * **PR #700** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/700" target="_blank">Graphite</a> 
          * **PR #701** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/701" target="_blank">Graphite</a> 
            * **PR #702** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/702" target="_blank">Graphite</a> 
              * **PR #703** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/703" target="_blank">Graphite</a> 
                * **PR #704** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/704" target="_blank">Graphite</a>  👈
        * **PR #689** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/689" target="_blank">Graphite</a> 
      * **PR #673** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/673" target="_blank">Graphite</a> 
      * **PR #672** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/672" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/deso-protocol/core/704?utm_source=stack-comment).

File: lib/pos_block_producer.go [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+
+func (pbp *PosBlockProducer) SignBlock(blockTemplate BlockTemplate, signerPrivateKey *bls.PrivateKey) (*MsgDeSoBlock, error) {
+	// TODO
+	return nil, nil
+}
+
+// CreateUnsignedBlock constructs an unsigned, PoS block with Fee-Time ordered transactions. This function should be used
+// during happy path in consensus when a vote QC has been assembled. The block is unsigned, so to indicate its incompleteness,
+// the block is returned as a BlockTemplate.
+func (pbp *PosBlockProducer) CreateUnsignedBlock(latestBlockView *UtxoView, newBlockHeight uint64, view uint64,
```
Comment (tholonious) [inline lib/pos_block_producer.go:48]:
```suggestion
func (pbp *PosBlockProducer) CreateUnsignedBlock(latestBlockView *UtxoView, newBlockHeight uint64, view uint64, proposerRandomSeedHash *RandomSeedHash,
```

File: lib/pos_block_producer.go [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+
+	// Fill out the validators vote QC field.
+	block.Header.ValidatorsVoteQC = validatorsVoteQC
+	return block, nil
+}
+
+// CreateUnsignedTimeoutBlock constructs an unsigned, PoS block with Fee-Time ordered transactions. This function should be used
+// during a timeout in consensus when a validators timeout aggregate QC has been assembled. The block is unsigned,
+// and so is returned as a BlockTemplate.
+func (pbp *PosBlockProducer) CreateUnsignedTimeoutBlock(latestBlockView *UtxoView, newBlockHeight uint64, view uint64,
```
Comment (tholonious) [inline lib/pos_block_producer.go:65]:
```suggestion
func (pbp *PosBlockProducer) CreateUnsignedTimeoutBlock(latestBlockView *UtxoView, newBlockHeight uint64, view uint64, proposerRandomSeedHash *RandomSeedHash,
```

File: lib/pos_block_producer.go [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+// and does not maintain any new internal state. Instead, most of the information needed to produce a block is passed in
+// as arguments. Both while instantiating the producer, or while creating a block to the CreateUnsignedBlock or
+// CreateUnsignedTimeoutBlock methods. As such, PosBlockProducer exists primarily for the purpose of cleaner separation of
+// concerns. Instantiating the PosBlockProducer can also be optional for nodes who do not wish to produce blocks.
+type PosBlockProducer struct {
+	mp                      Mempool
+	params                  *DeSoParams
+	proposerPublicKey       *PublicKey
+	proposerVotingPublicKey *bls.PublicKey
+	proposerRandomSeedHash  *RandomSeedHash
```
Comment (tholonious) [inline lib/pos_block_producer.go:26]:
```suggestion
```

**Review by tholonious — CHANGES_REQUESTED** [review]
LGTM aside from the propposerRandomSeedHash param change

**Review by tholonious — APPROVED** [review]

**Pull request merged**